man/socket_setup.3
.TH qlibs: socket_setup 3
.SH NAME
socket_setup \- listen to or accept socket for incoming TCP connections
.SH SYNTAX
.B #include \(dqsocket_if.h\(dq

int \fBsocket_listen\fP(int \fIs\fR,int \fIn\fR);

int \fBsocket_accept\fP(int \fIs\fR,char \fIip\fR[16],
                  uint16 *\fIport\fR,uint32 *\fIscope_id\fR);
                 
int \fBsocket_ipoptionskill\fR(int \fIs\fR);

int \fBsocket_ip6anycast(\fR(int \fIs\fR);

int \fBsocket_dualstack\fR(int \fIs\fR);

int \fBsocket_nodualstack\fR(int \fIs\fR);
.SH DESCRIPTION
.B socket_listen 
prepares TCP socket \fIs\fR to accept TCP connections.  
It allows a backlog of approximately \fIn\fR TCP SYNs. 
(On systems supporting SYN cookies, the backlog is irrelevant.) 

.B socket_accept 
accepts the connection. It creates a new socket
for the connection and returns a file descriptor pointing to the new
socket; you can use the read and write system calls to transmit data
through that file descriptor.
Further, it provides information about client's 
\fIip\fR address and TCP \fIport\fR number
perhaps together with local receiving interface \fIscope_id\fR. 

.B socket_ipoptionskill
is used to disable previously defined options in IPv4 or IPv6 packets 
like Source Routing prior of using this socket for data exchange.
.B socket_ipoptionskill
uses the 
.BR setsockopt .

.B socket_ip6anycast
enables unspecified reversed anycasting on the listening socket
.IR s 
with IPv6 address
.IR :: .
Upon receiving IPv6 packets, the socket records the
incoming IPv6 address and the receiving \fIscope_id\fR
in order provide additional routing information.

.B socket_dualstack
and
.B socket_nodualstack
can be used to force or forbid dual-stack behavior
setting the 
.B setsockopt 
variable
.I IPV6_V6ONLY 
appropriately. In the last case, a potential servers
needs two instances to accept incoming IPv6 and IPv6 packets.
.SH "RETURN CODES"
Normally
.BR socket_listen ,
.B socket_accept 
and
.B socket_ipotionskill 
as well as 
.B socket_dualstack
and
.B socket_nodualstack
return
.I 0 
and if anything goes wrong it returns 
.IR -1 , 
setting 
.I errno
appropriately.
.SH EXAMPLE
  #include <socket_if.h>

  int \fIs\fR, \fIt\fR; 
  int \fIr\fR;
  char \fIip\fR[16];
  uint16 \fIp\fR;

  if ((s = socket_tcp()) == -1)
    strerr_die2sys(111,FATAL,"unable to create TCP socket: ");
  r = socket_ipoptionskill(s);
  if (socket_bind_reuse(s,(char *)V6localnet,8002,0) == -1)
    strerr_die2sys(111,FATAL,"unable to bind: ");
  if (socket_listen(s,1) == -1)
    strerr_die2sys(111,FATAL,"unable to listen: ");
    
  t = socket_tcp();
  socket_bind(t,ip,p,0);
  socket_listen(s,16);
  socket_accept(t,ip,&p,&scope_id);  
.SH "SEE ALSO"
socket_if(3), 
socket_bind(3), 
socket_connect(3), 
socket_info(3),
socket_recv(3), 
socket_send(3), 
socket_tcp(3), 
socket_udp(3)


CHANGELOG
fehQlibs-09:	First public release based on Qlibs as 
		a common project of Kai Peter and Erwin Hoffmann.
fehQlibs-10:	Added IPV6_V6ONLY setsockoptions call.
		Improved poll handling in iopause.c considering poll RC.
		Added building of shared libraries + updated docs.
fehQlibs-11:	Added ia4_fmt, ia6_fmt and dns_cname (for *qmail).
		Changed dns_transmit lookup time constants. 
fehQlibs-12:	dns_ip, dns_cname, dns_txt, dns_name and dns_mx
		return now the number of DNS answers received 
		unlike the number of bytes for the given output. 
		Added convenience routines for forthcoming s/qmail.
fehQlibs-12a:	Fixed missing DNS A lookup in dns_ip.c.
fehQlibs-12b:	Removed (one) obsolete return value check 
		for iopause in timeoutconn.c.
fehQlibs-12c:	Checking carefully all dnsstub return codes
		while using a coherent scheme. 
fehQlibs-13:	Bugs fixed: str_copy (wrong paranthesis setting)
		UI changed:	log() -> log_who()
		Added: man str(3)
fehQlibs-13a:	Fixed wrong input boundary checking for multiple
	DNSCACHE variables settings in dns_rcip.c. (Tx. J.W.)
fehQlibs-13b:	Added missing str_copyb() function in str.c.
fehQlibs-13c:	iopause it RC < 0 forces end of connection loop.
	DNS stub return codes straightend with djbdnscurve6-36.
fehQlibs-13d:	Fixed integration bugs not displaying IPv6 addresses
		(dns_ip.c, dns_mx.c).
		Wrong dependency check in Makefile for SHAREDLIBS (tx. Alan S.). 
fehQlibs-14:	Added 'const' for most input arguments + updated man:
		byte, logmsg, str, pathexec. Patch contributed by Alan S.
		Added byte_fill and case_upper function (potentially for SRS).
		Included CLFAGS and LDFLAGS macros in Makefile.


socket_setup.c
#include <sys/types.h>
#include <sys/param.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include "byte.h"
#include "socket_if.h"
#include "ip.h"

/**
  @file socket_setup.c
  @author djb, feh
  @source ucspi-tcp6
  @brief setup listening socket
*/


int socket_accept(int s,char ip[16],uint16 *port,uint32 *scope_id)
{
  struct sockaddr_in6 sa;
  unsigned int dummy = sizeof(sa);
  int fd;

  fd = accept(s,(struct sockaddr *)&sa,&dummy);
  if (fd == -1) return -1;

  if (sa.sin6_family == AF_INET) {
    struct sockaddr_in *sa4 = (struct sockaddr_in*)&sa;
    byte_copy(ip,12,V4mappedprefix);
    byte_copy(ip+12,4,(char *)&sa4->sin_addr);
    uint16_unpack_big((char *)&sa4->sin_port,port);
    if (scope_id) *scope_id = 0;
    return fd;
  }
  byte_copy(ip,16,(char *)&sa.sin6_addr);
  uint16_unpack_big((char *)&sa.sin6_port,port);
  if (scope_id) *scope_id = sa.sin6_scope_id;

  return fd;
}

int socket_listen(int s,int backlog)
{
  return listen(s,backlog);
}

int socket_ipoptionskill(int s)
{
  if (ipv4socket) 
    return setsockopt(s,IPPROTO_IP,1,(char *) 0,0); /* 1 == IP_OPTIONS */

  return setsockopt(s,IPPROTO_IPV6,1,(char *) 0,0); 
}

int socket_ip6anycast(int s)
{
  int opt = 1;
  int r;

#ifdef GEN_IP_PKTINFO /* Linux */
   r = setsockopt(s,IPPROTO_IP,GEN_IP_PKTINFO,&opt,sizeof(opt));
#elif IP_PKTINFO  /* Solaris */
   r = setsockopt(s,IPPROTO_IP,IP_PKTINFO,&opt,sizeof(opt));
#elif IP_RECVDSTADDR  /* BSD */
   r =  setsockopt(s,IPPROTO_IP,IP_RECVDSTADDR,&opt,sizeof(opt));
#elif IPV6_RECVDSTADDR
   if (!ipv4socket) setsockopt(s,IPPROTO_IPV6,IPV6_ONLY,&opt,sizeof(opt));
   r = setsockopt(s,IPPROTO_IPV6,IP_RECVDSTADDR,&opt,sizeof(opt));
#endif
   return r;
} 

int socket_dualstack(int s)
{
  int opt = 0;

  return setsockopt(s,IPPROTO_IPV6,IPV6_V6ONLY,&opt,sizeof(opt));
}

int socket_nodualstack(int s)
{
  int opt = 1;

  return setsockopt(s,IPPROTO_IPV6,IPV6_V6ONLY,&opt,sizeof(opt));
}


